openpype/pipeline/colorspace.py

```diff
@@ -0,0 +1,464 @@
+    project_settings=None,
+    validate=True
+):
+    """Get colorspace name from filepath
+
+    ImageIO Settings file rules are tested for matching rule.
+
+    Args:
+        path (str): path string, file rule pattern is tested on it
+        host_name (str): host name
+        project_name (str): project name
+        config_data (dict, optional): config path and template in dict.
+                                      Defaults to None.
+        file_rules (dict, optional): file rule data from settings.
+                                     Defaults to None.
+        project_settings (dict, optional): project settings. Defaults to None.
+        validate (bool, optional): should resulting colorspace be validated
+                                   with config file? Defaults to True.
+
+    Returns:
+        str: name of colorspace
+    """
+    if not any([config_data, file_rules]):
+        project_settings = project_settings or get_project_settings(
+            project_name
+        )
+        config_data = get_imageio_config(
+            project_name, host_name, project_settings)
+        file_rules = get_imageio_file_rules(
+            project_name, host_name, project_settings)
+
+    # match file rule from path
+    colorspace_name = None
+    for _frule_name, file_rule in file_rules.items():
+        pattern = file_rule["pattern"]
+        extension = file_rule["ext"]
+        ext_match = re.match(
+            r".*(?=.{})".format(extension), path
+        )
+        file_match = re.search(
+            pattern, path
+        )
+
+        if ext_match and file_match:
+            colorspace_name = file_rule["colorspace"]
+
+    if not colorspace_name:
+        log.info("No imageio file rule matched input path: '{}'".format(
+            path
+        ))
+        return None
+
+    # validate matching colorspace with config
+    if validate and config_data:
+        validate_imageio_colorspace_in_config(
+            config_data["path"], colorspace_name)
+
+    return colorspace_name
+
+
+def parse_colorspace_from_filepath(
+    path, host_name, project_name,
+    config_data=None,
+    project_settings=None
+):
+    """Parse colorspace name from filepath
+
+    An input path can have colorspace name used as part of name
+    or as folder name.
+
+    Args:
+        path (str): path string
+        host_name (str): host name
+        project_name (str): project name
+        config_data (dict, optional): config path and template in dict.
+                                      Defaults to None.
+        project_settings (dict, optional): project settings. Defaults to None.
+
+    Returns:
+        str: name of colorspace
+    """
+    if not config_data:
+        project_settings = project_settings or get_project_settings(
+            project_name
+        )
+        config_data = get_imageio_config(
+            project_name, host_name, project_settings)
+
+    config_path = config_data["path"]
```

Comment:
Would this function make more sense when simplified/rewritten as just finding the first relevant color space from the list of color spaces of the config? It'd make it more versatile in use instead of it internally always needing to call `get_ocio_config_colorspaces`.

It feels to me that parsing from filenames like this might often be frequently done close after each other - e.g. parsing multiple files. Quering the color space files rules will be relatively slow since it requires accessing the OCIO config file each time.

```python
colorspaces = get_ocio_config_colorspaces(config_path)
for filepath in filepaths:
    colorspace = parse_colorspace_from_filepath(filepath, colorspaces=colorspaces)
```

This also looks to me like a much more friendly API to the user with much less arguments to pass. Plus it'd be much faster and a simpler function code body.

As a general recommendation I'd try to see if there are other areas where functions could be simplified like this. Needing to pass project name, host, project settings to all functions where they end up just parsing an OCIO config path I'd say should be rewritten to just take the config path.

Reply:
This comment wasn't resolved, was it?

Reply:
Sorry, yeah my bad. I had converted it into issue https://github.com/ynput/OpenPype/issues/4348 but the automation is not adding the reference within this comment.